### ISSUE TITLE
GPU dispatch architecture: async completion (MSI-X or polling thread) for batched launches

### ISSUE THREAD

**Issue by johnjezl** [SLM-OS/SLM-Operating-System#573]
## Background

SLM-OS's current GA10B dispatch path (`kernel/gpu/nvidia/ga10b_bringup.c:1482` `ga10b_submit_and_poll`) is **synchronous per launch**: build pushbuffer for one op → write GPFIFO entry → ring doorbell → busy-poll a per-op semaphore until it lands → repeat. This worked well enough for MNIST's 8 ops/inference but doesn't scale to real SLM workloads.

Estimate for Qwen 2.5 1.5B (the target model per ongoing M5.3.x work):
- ~370 ops per forward pass (28 transformer layers × ~13 ops/layer + embedding + LM head + sampler)
- At a modest 10 tokens/sec generation rate, that's ~3700 launches/sec

Per-launch CPU/IO overhead in `submit_and_poll`:
- QMD build/copy + cache_clean: ~5 µs
- Pushbuffer write + GPFIFO entry + USERD GP_PUT update + doorbell: ~10 µs
- Busy-poll semaphore loop: 10-50 µs depending on op duration

**~30-50 µs of pure dispatch-side overhead per launch**, on top of the GPU's actual compute. At 3700 launches/sec that's **~120-180 ms/sec of CPU spent on dispatch overhead alone**, scaling linearly with token rate. This pegs a CPU core for what should be GPU-bound work.

## What production drivers do

CUDA runtime, NVK (Mesa Vulkan), nouveau, and HailoRT all use the same general pattern:

1. **Batch many launches into one pushbuffer.** Pushbuffer writes are cheap; PBDMA processes a burst of GPFIFO entries faster than CPU can submit them one-at-a-time.
2. **Fire one doorbell per batch**, not one per launch.
3. **Async completion signaling.** Either MSI-X interrupts on per-batch fences, or one polling thread per channel rather than per-launch polling. The dispatch thread queues work and moves on.
4. **Channel runs ahead of CPU.** GPU keeps draining the GPFIFO; CPU keeps refilling it. Steady-state CPU dispatch overhead is amortized across many ops.

Reference: `../slmos-reference-cache/mesa/mesa-nvk_queue.c` and `mesa-nvk_cmd_dispatch.c` for the NVK pattern, `../slmos-reference-cache/nvgpu-common-fifo-submit.c` for L4T's pattern.

## Why this matters now

Two things converging make this the next bottleneck after the off-by-one fix lands:

1. **Per-token op count grows ~50× from MNIST to Qwen.** The synchronous overhead was negligible at 8 ops/inference; it becomes the dominant cost at 370 ops/inference.
2. **The fix for #558 (per-dispatch QMD construction) adds ~5 µs per launch on top.** Tolerable in absolute terms, but it's another constant added to a hot path that's already expensive per-launch.

## What's out of scope for this issue

- The off-by-one bug itself — that's #558, currently being fixed via per-dispatch QMD construction (Option 3 in `docs/gpu-qmd-per-dispatch-plan.md`). The work tracked here is the *next* layer once correctness is solved.
- IRQ-off wedge under burst load (#535) — different failure mode. Async dispatch may incidentally help, but not the primary motivation.

## Suggested phases

1. **Pushbuffer batching.** Rewrite `submit_and_poll` to build multi-launch pushbuffers (N ops at a time) and fire one doorbell per batch. Still synchronous-per-batch, but batch size of 8-32 amortizes most of the per-launch overhead. No completion-signaling changes yet.
2. **Async completion via MSI / SGI.** Move semaphore polling off the dispatch thread. Dispatch thread queues batches and returns immediately; a separate completion path (interrupt-driven if the GA10B IRQ delivery path supports it post-kexec, or a single dedicated polling thread otherwise) signals when a batch's last semaphore lands.
3. **Pipeline parallelism.** Multiple in-flight batches per channel; the GPU drains the GPFIFO continuously while CPU stays ahead. This is where the real throughput gains live.
4. **Multi-channel for SLM concurrency.** Once a single channel is fully utilized, additional channels for serving multiple inference sessions concurrently.

Phases 1-2 are likely worth doing for the capstone Qwen demo. Phases 3-4 are post-capstone.

## Acceptance criteria (Phase 1, the immediate next step)

- Qwen 1.5B forward pass on GA10B with per-dispatch QMD construction shows token-rate proportional to GPU compute time, not CPU dispatch overhead.
- Concrete metric: ratio of "CPU time spent in `ga10b_*` functions" to "wall time" should be < 20% at sustained inference rate. Today it would be approaching 100% if we drove Qwen through the synchronous path.

## References

- `kernel/gpu/nvidia/ga10b_bringup.c:1482` — current `submit_and_poll`
- `docs/gpu-qmd-per-dispatch-plan.md` §7 — context where this overhead was first characterized
- Issue #558 — the correctness bug being fixed first
- Issue #535 — IRQ-off wedge (related but different failure mode)

**Comment by johnjezl** [comment]
## Status update — PR #612 partially closes this

PR #612 (merged 2026-05-04) refactored `ga10b_dispatch_v7_pipeline` to nvgpu's bulk-submit pattern. That delivers two of the four "what production drivers do" items from the issue body:

- ✅ **Batch many launches into one pushbuffer.** All N op pushbuffers are packed into the contiguous pushbuf area (op[0]..op[N-1] launch_kernel pb's + 1 trailing COMPUTE_B sema-release pb).
- ✅ **Fire one doorbell per batch.** USERD GP_PUT advances once to the cumulative new value, doorbell rings once for the whole chain.

Empirical: 200-iter alternating-fill MNIST probe drops from ~360 sec (per-op + band-aid) to **65 ms** with `m=0/200` mismatches. ~5500× speedup on this workload.

## What's still open under this title

- ❌ **Async completion signaling.** The bulk dispatcher still busy-polls the trailing sema for up to 2 s. No MSI-X interrupt path, no per-channel polling thread. The dispatch CPU is blocked on the chain's completion just like before — just with one poll instead of N.
- ❌ **Channel runs ahead of CPU.** Only one batch is in flight at a time; the next chain waits for the previous to drain. No pipelining of multiple chains.

For Qwen at 10 tok/s × ~370 ops/token = 3700 launches/sec, the bulk pattern reduces dispatch overhead from ~30-50 µs/launch to whatever the per-batch sync cost amortizes to. Per-batch sync is still synchronous so there's a per-batch tail-latency floor; the issue's headline of "amortizing CPU dispatch overhead across many ops" is now bounded by batch granularity rather than per-op.

## Suggested next-action shape

Re-title/re-scope this issue to focus on the remaining piece:

> "GPU dispatch architecture: async completion (MSI-X or polling thread) for batched launches"

The "batched launch" half landed via PR #612 / #596. The async/multi-chain half is what unblocks scaling past per-batch sync limits.

**Comment by johnjezl** [comment]
## Option 1 attempt — UNSAFE under current lock structure

Tried the simplest path proposed earlier (cooperative yield in the existing poll loop) on branch [`feat/ga10b-poll-yield`](https://github.com/SLM-OS/SLM-Operating-System/tree/feat/ga10b-poll-yield) (commit 4a7d3f2c, **NOT merged**, branch kept as a record of the failed experiment).

### Result on jetson-nano-2

Same 200-iter alternating-fill probe:

| Build | Mismatches | Total runtime |
|---|---|---|
| PR #612 (busy-poll, baseline) | 0 / 200 | 65 ms |
| Option 1 (spin + yield) | **118 / 200** | 139 ms |

Both regressions: 59% wrong-argmax rate AND 2× slowdown.

### Root cause

`slm_gpu_run_mnist` in `kernel/src/slm_ffi.c:639-661` holds `g_gpu_dispatch_lock` via `spin_lock_irqsave` across the **entire** `ga10b_bringup_launch_kernel` call — including the dispatcher's poll loop. So my `yield()` ran with **interrupts disabled and a spinlock held**.

That is the explicit CLAUDE.md "Critical" anti-pattern:

> "yield() or sleep() called with interrupts disabled"

Yielding under those conditions causes:
- The new task running may try to acquire the same dispatch lock (block, but with confused IRQ state)
- The IRQ state on the resuming CPU is wrong relative to the scheduler's expectations
- Subtle cross-task corruption of GPU dispatch state, manifesting as wrong logits

The 59% mismatch rate is consistent with channel-state inconsistency every other iteration or so.

### Implication for this issue

Option 1 (yield in the existing poll loop, same calling task) cannot work under the current lock structure. The fix would require one of:

1. **Move the lock UP to the caller** so `ga10b_bringup_launch_kernel` runs lock-free. But callers expect serialized GPU dispatch and would need a new synchronization primitive.
2. **Release the lock during the poll, re-acquire after**. Allows concurrent dispatches that race on channel state.
3. **A separate polling task that doesn't hold the dispatch lock** — Option 2/3 from the original triage. The polling task pops requests off a queue (already lock-protected), runs the dispatch, polls without holding `g_gpu_dispatch_lock`, signals the original caller via wait-handle.

Option 3 is the architecturally clean form and matches the issue's retitle. It's substantial work (~150-400 lines depending on whether multi-chain pipelining is in scope) but is the only safe path to "CPU not pinned during GPU compute."

### Out of scope for #612 / safe-to-keep state

PR #612's busy-poll behavior is the safe baseline. No change to current main. This issue stays open with the polling-task work as the remaining scope; ditched the yield-in-poll experiment.

**Comment by johnjezl** [comment]
## Option 2 attempt — works structurally, has cold-start regression

Branch [\`feat/gpu-dispatcher-task\`](https://github.com/SLM-OS/SLM-Operating-System/tree/feat/gpu-dispatcher-task) (commit \`b06870ef\`, **NOT merged**, kept as record of the implementation).

### Design

- One \`gpu_dispatcher\` task spawned at boot from \`kernel_main\` (post \`ipc_init\`, pre \`scheduler_start\`).
- Two queues: \`g_gpu_request_q\` (capacity 8) for caller→dispatcher, plus a per-call reply queue created on each \`gpu_dispatch_via_task\` invocation.
- Caller's \`slm_gpu_run\` / \`set_input\` / \`set_input_fill\` / \`run_with_input\` build a request struct on the caller stack, msg_send it, msg_recv on the per-call reply queue.
- Dispatcher's task body: msg_recv from request queue → take \`g_gpu_dispatch_lock\` → switch on op type → run the actual \`ga10b_bringup_*\` work → release lock → msg_send reply.
- Lock retained to mutex against the \`nvgpu\` shell-command path.

Builds clean for Jetson + QEMU. \`make test\` and \`make test-ga10b-bringup\` both green.

### Hardware result

Single dispatch call: works.
N=10 alternating-fill probe on jetson-nano-2:

\`\`\`
iter 1 input=0 argmax=5  ← WRONG (expected 6)
iter 2 input=1 argmax=5  ← right
iter 3 input=0 argmax=6  ← right
iter 4 input=1 argmax=5  ← right
... all subsequent correct
\`\`\`

So 1/200 mismatch rate (vs PR #612 baseline of 0/200). Single iter regression on cold start.

### Diagnosis

The dispatcher invokes the existing \`ensure_bringup\` Bug-A warmup on the FIRST request (a \`set_input_fill\` from the probe). The warmup runs \`ga10b_bringup_launch_kernel\` BEFORE the user's set_input takes effect, so the warmup's output / GPU state somehow leaks into iter 1's user-visible result.

Why this doesn't happen on the direct path (PR #612 measured 0/200 on the same probe shape) is the open mystery — the call sequence is structurally identical. Possible differences:
- Task identity (caller-task vs dispatcher-task) affecting some context-sensitive state I haven't found
- Subtle interaction between the per-call reply queue alloc/free and PMM state
- A race in msg_send/msg_recv tear-down that wasn't visible at the lock-held call shape

### What this means for #573

The polling-task design is structurally correct and unlocks the original goal — caller CPUs are free during GPU compute, no IRQ-off-yield anti-pattern. But the cold-start regression makes it not mergeable as-is.

A future iteration would need to either:
1. Skip the warmup in \`ensure_bringup\` (let iter 1 take the Bug A hit, unchanged from before warmup landed)
2. Move the warmup to a post-\`scheduler_start\` background dispatch (so it's done by the time the first user request arrives)
3. Find the actual context dependency that makes iter 1 wrong on the dispatcher path but not the direct path

Option 2 is the cleanest — eliminates the lock-hold-time tradeoff that motivated the warmup comment in the first place.

For now, the branch sits as a code-reusable starting point. Closing the implementation loop on this issue is deferred.
